### PR TITLE
Add remove page button to top toolbar

### LINKED ISSUE #3279: Simple way to remove AMP story pages
## Feature description

As a user I want there to be an easier way to remove pages from within the editor.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* AC1: It is currently possible to remove AMP story pages by pressing delete or using the remove block button on the in the top bar. However, there should be a simple way to remove pages within the UI.
* AC2: Add "Remove page" to the right-click context menu, upon selection, page will be removed.
* AC3: Include a "remove" link in the page panel, when a page is selected. Selecting the "remove" link will remove the page (see visual in comment below).

## Implementation brief
Implementation for this issue is broken into two piece of work. 

1. Add another a Remove page back on the `InspectorControls` for the edit method the page block. 
Mock up of which looks like this. 

Screenshot from 2019-10-15 16-17-30

2. Add new options to the context menu that appears when a user right clicks in the stories editor. 
Currently the context menu is limiting to only appearing in one of two cases. First if a block other the the page block is selected, it shows options. Case two, if a page block is selected and there a block in the clipboard, show a 'Paste block' option. 

To implement this feature, some of the limitations and checks for if page is selected. These will have to be removed. The following limitations will need to be respected. 

- When a page block is selected, a user will not be able to copy or cut or page a page. 
- However, they is no limitation on duplication, so if simple, this should also be implemented. 
- Elements in popover menu, should say, remove page instead of remove block. 
- If only one page is in the story, remove page should not be displayed. 

A mockup of what this will look like. 
Screenshot from 2019-10-15 15-40-31

### Test 1
- Create new story post. 
- Create a second page. 
- Select second page. 
- Click page to show right context menu (background settings). 
- Click remove page. 
- Confirm page is removed.   

### Test 2
- Create new story post. 
- Create a second page. 
- Select second page. 
- Right Click page to show context menu
- Click remove page. 
- Confirm page is removed.  

## Demo

* https://youtu.be/IyWXV3H9ysc
* https://youtu.be/ncaN-N2dLfA

## Changelog entry

* Added a new button to the sidebar to easily remove pages.

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary
Add remove page button to top toolbar

See #3279 

## Checklist

- [ ] My pull request is addressing an [open issue](https://github.com/ampproject/amp-wp/contributing/project-management.md#life-of-an-issue) (please create one otherwise).
- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/contributing/engineering.md#tests).
- [ ] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/contributing/engineering.md) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

**Comment by spacedmonkey** [comment]
I could also do with some feedback on testing here.

File: assets/src/stories-editor/plugins/context-menu-item.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+		<PluginBlockSettingsMenuItem
+			icon="trash"
+			label={ __( 'Remove page', 'amp' ) }
+			role="menuitem"
+			onClick={ removePage }
+		/>
+	);
+};
+
+export const render = () => {
```
Comment (barklund) [inline assets/src/stories-editor/plugins/context-menu-item.js]: This is similar to simply writing:

```
export const render = () => <RemovePageSetting />;
```

File: assets/src/stories-editor/plugins/context-menu-item.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ * Renders remove page button.
+ *
+ * @return {ReactElement} The rendered trash menu item.
+ */
+const RemovePageSetting = () => {
+	const { removeBlock } = useDispatch( 'core/block-editor' );
+	const currentPageId = useSelect( ( select ) => select( 'amp/story' ).getCurrentPage(), [] );
+
+	const removePage = () => {
+		if ( ! isPageBlock( currentPageId ) ) {
```
Comment (barklund) [inline assets/src/stories-editor/plugins/context-menu-item.js]: When will `getCurrentPage()` ever return something that isn't a page?

Comment (swissspidy) [reply assets/src/stories-editor/plugins/context-menu-item.js]: Good question. I don't think there's ever such a case. So this callback could be simplified.

Could probably even use `useCallback` to get a memoized callback:

```js
const removePage = useCallback(
  () => removeBlock( currentPageId ),
  [ currentPageId, removeBlock ],
);
```

File: assets/src/stories-editor/plugins/context-menu-item.js (hunk 1 above)
Comment (spacedmonkey) [reply assets/src/stories-editor/plugins/context-menu-item.js]: Good spot. It was like after a refactor. Changed in https://github.com/ampproject/amp-wp/pull/3538/commits/5a65afcebe406c461f7fc134877f6b101f5d08da

File: assets/src/stories-editor/plugins/context-menu-item.js (hunk 2 above)
Comment (spacedmonkey) [reply assets/src/stories-editor/plugins/context-menu-item.js]: I think I was being overly careful here. I have removed the `isPageBlock` check. 

Not sure on the benefit of `useCallback` but I have put in anyway. https://github.com/ampproject/amp-wp/pull/3538/commits/5a65afcebe406c461f7fc134877f6b101f5d08da

File: assets/src/stories-editor/plugins/context-menu-item.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
+/**
+ * Plugin for add remove page button to more settings.
+ */
+
+/**
+ * External dependencies
+ */
+import { ReactElement } from 'react';
+/**
```
Comment (swissspidy) [inline assets/src/stories-editor/plugins/context-menu-item.js]:
```suggestion

/**
```

File: assets/src/stories-editor/plugins/context-menu-item.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ * Plugin for add remove page button to more settings.
```
Comment (swissspidy) [inline assets/src/stories-editor/plugins/context-menu-item.js]:
```suggestion
 * Plugin that adds a remove page button to the "More" settings menu.
```

File: assets/src/stories-editor/plugins/context-menu-item.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
+/**
+ * Plugin that adds a remove page button to the "More" settings menu.
+ */
+
+/**
+ * External dependencies
+ */
+import { ReactElement } from 'react';
```
Comment (barklund) [inline assets/src/stories-editor/plugins/context-menu-item.js]: Is this used? And if it is, shouldn't it be an import from `@wordpress/element`?

File: assets/src/stories-editor/plugins/context-menu-item.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+	const removePage = useCallback(
+		() => removeBlock( currentPageId ),
+		[ currentPageId, removeBlock ],
+	);
+
+	return (
+		<PluginBlockSettingsMenuItem
+			icon="trash"
+			label={ __( 'Remove page', 'amp' ) }
```
Comment (swissspidy) [inline assets/src/stories-editor/plugins/context-menu-item.js]:
```suggestion
			label={ __( 'Remove Page', 'amp' ) }
```

File: assets/src/stories-editor/plugins/context-menu-item.js (hunk 5 above)
Comment (spacedmonkey) [reply assets/src/stories-editor/plugins/context-menu-item.js]: That was a copy and paste from somewhere else in the code base. I don't love importing but I got a warning in the lints if I didn't have it.

**Comment by spacedmonkey** [comment]
> This menu item should not be rendered when `selectedBlock === currentPageId`. Very confusing otherwise.

This is a just confusing behaviour. If means you have to select another block to be able to remove the page. There is no simple way to select a page. So it is unclear to a user when this button would appear.

File: assets/src/stories-editor/plugins/context-menu-item.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	const currentPageId = useSelect( ( select ) => select( 'amp/story' ).getCurrentPage(), [] );
+	const pages = useSelect( ( select ) => select( 'core/block-editor' ).getBlockOrder(), [] );
+
+	const removePage = useCallback(
+		() => removeBlock( currentPageId ),
+		[ currentPageId, removeBlock ],
+	);
+
+	// Shouldn't allow users to remove the first and only page.
+	if ( pages.length < 2 ) {
```
[the remark this answers is not in the document]
Comment (spacedmonkey) [reply assets/src/stories-editor/plugins/context-menu-item.js]: What is the point of a fresh start? Why not just create a new story at the point....

**Comment by spacedmonkey** [comment]
Any changes of behaviour should discussed on the original ticket as they are changes to the IB.

File: assets/src/stories-editor/plugins/context-menu-item.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (miina) [reply assets/src/stories-editor/plugins/context-menu-item.js]: (It's actually not more clicks if I haven't written the title yet, I take that back :) )

**Comment by spacedmonkey** [comment]
I am closing PR, after feedback on #3279 . We should focus on #3466 .

**Pull request closed without merging**